InterruptClient.rb

```diff
@@ -0,0 +1,230 @@
+require 'socket'
+require 'json'
+
+require_relative './Console'
+
+class InterruptClient
+
+  CMD_QUIT = "\\"
+  START_MSG = 'Welcome! What is your name?'
+  INSTRUCTIONS = "Start typing to join the chat! To quit anytime, type #{CMD_QUIT}"
+  FAREWELL = "\r\nbye"
+  PROMPT = '> '
+
+  MAX_MSG_LENGTH = 3000 # max length of incoming message read
+  HANDSHAKE_WAIT = 2 # number of seconds to wait for ack from server before resending
+
+  COLORS = [
+    'green',
+    'magenta',
+    'cyan',
+    'blue',
+    'light_green',
+  ]
+
+  def initialize(server_host, server_port)
+    @server_host = server_host
+    @server_port = server_port
+
+    @client = UDPSocket.new
+    @client.connect(@server_host, @server_port)
+
+    @latest_chat = '' # string timestamp of latest chat received
+  end
+
+  def run
+    set_name
+    handshake
+    instructions
+    terminal_config
+    receive_loop
+  end
+
+  def set_name
+    puts START_MSG
+    print PROMPT
+    @name = STDIN.gets.chomp
+  end
+
+  def handshake
+    ackd = false
+    while not ackd
+      send_msg(msg_connect(@name))
+      incoming = IO.select([@client], nil, nil, HANDSHAKE_WAIT)
+      if ((not incoming.nil?) && ack?)
+        ackd = true
+      end
+    end
+  end
+
+  def ack?
+    data, sender = @client.recvfrom(MAX_MSG_LENGTH)
+    msg = parse_msg(data, sender)
+    return false if msg.nil?
+
+    msg['type'] == 'ack'
+  end
+
+  def instructions
+    print Console.clear
+    puts INSTRUCTIONS + "\n" * 12
+  end
+
+  def receive_loop
+    loop do
+      to_read = IO.select([@client, STDIN]) # check what's ready to read: socket, or terminal input
```

Comment:
looks like you can do

```ruby
client_socket, stdin_socket = IO.select([@client, STDIN])
```
http://ruby-doc.org/core-2.0.0/IO.html#method-c-select

that should save you from having to figure out where the IO select came from